### PR TITLE
Refactor FixedPool to remove template parameters

### PR DESCRIPTION
**Pull request by jdahm**

### CODE REVIEW

File: src/umpire/strategy/FixedPool.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+  return ptr;
+}
+
+void
+FixedPool::deallocate(void* ptr)
+{
+  for (auto& p : m_pool) {
+    const int object_index = (reinterpret_cast<char*>(ptr) - p.data) / m_obj_bytes;
+    if (static_cast<unsigned int>(object_index) < m_obj_per_pool) {
```
Comment (jdahm) [inline src/umpire/strategy/FixedPool.cpp]: I'm not quite sure this works in all cases yet. I'm counting on the sign change causing a large enough index to fail the check.

**Comment by jdahm** [comment]
This is ready for review!

**Comment by mcfadden8** [comment]
Thanks @jdahm, are there any performance differences between this implementation and the template version that is in tpl?

**Comment by jdahm** [comment]
On my mac, with system noise:

New
```
FixedPoolHost/allocate/4          364 ns        364 ns    1920887
FixedPoolHost/deallocate/4        514 ns        514 ns    1299642
```

Old
```
FixedPoolHost/allocate/4          413 ns        412 ns    1745270
FixedPoolHost/deallocate/4        472 ns        470 ns    1500549
```

File: src/umpire/strategy/FixedPool.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+FixedPool::Pool::Pool(AllocationStrategy* allocation_strategy,
+                      const size_t object_bytes, const size_t objects_per_pool,
+                      const size_t avail_bytes) :
+  strategy(allocation_strategy),
+  data(reinterpret_cast<char*>(strategy->allocate(object_bytes * objects_per_pool))),
+  avail(reinterpret_cast<int*>(std::malloc(avail_bytes))),
+  num_avail(objects_per_pool)
+{
+  const unsigned char not_zero = ~0;
```
Comment (davidbeckingsale) [inline src/umpire/strategy/FixedPool.cpp]: This sets everything to 1? I think anything involving a bit twiddle should have a comment on what is happening and why.

File: src/umpire/strategy/FixedPool.cpp [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+FixedPool::allocInPool(Pool& p) noexcept
+{
+  if (!p.num_avail) return nullptr;
+
+  for (unsigned int int_index = 0; int_index < m_avail_length; ++int_index) {
+    const int bit_index = ffs(p.avail[int_index]) - 1;
+    if (bit_index >= 0) {
+      const size_t index = int_index * bits_per_int + bit_index;
+      if (index < m_obj_per_pool) {
+        p.avail[int_index] ^= 1 << bit_index;
```
Comment (davidbeckingsale) [inline src/umpire/strategy/FixedPool.cpp]: Add a comment explaining what's going on here

File: src/umpire/strategy/FixedPool.cpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    const char* t_ptr = reinterpret_cast<char*>(ptr);
+    const ptrdiff_t offset = t_ptr - p.data;
+    if ((offset >= 0) && (offset < static_cast<ptrdiff_t>(m_data_bytes))) {
+      const size_t alloc_index = offset / m_obj_bytes;
+      const size_t int_index   = alloc_index / bits_per_int;
+      const short  bit_index   = alloc_index % bits_per_int;
+
+      UMPIRE_ASSERT(! (p.avail[int_index] & (1 << bit_index)));
+
+      p.avail[int_index] ^= 1 << bit_index;
```
Comment (davidbeckingsale) [inline src/umpire/strategy/FixedPool.cpp]: Comment please.

File: src/umpire/strategy/FixedPool.cpp [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  m_pool.emplace_back(m_strategy, m_obj_bytes, m_obj_per_pool, m_avail_length * sizeof(int));
+}
+
+void*
+FixedPool::allocInPool(Pool& p) noexcept
+{
+  if (!p.num_avail) return nullptr;
+
+  for (unsigned int int_index = 0; int_index < m_avail_length; ++int_index) {
+    const int bit_index = ffs(p.avail[int_index]) - 1;
```
Comment (davidbeckingsale) [inline src/umpire/strategy/FixedPool.cpp]: Comment here.

**Comment by davidbeckingsale** [comment]
@jdahm - this looks good, can you get it up-to-date with `develop`?

**Comment by davidbeckingsale** [comment]
Let's merge this once the tests pass!

**Pull request merged**